File: compressor/format.py
import struct
from typing import Iterable, Iterator

from compressor.block import Block, BlockMetadata

MAGIC = b"NOVA"
VERSION = 1
HEADER_STRUCT = struct.Struct("<4sH I 32s")
BLOCK_HEADER_STRUCT = struct.Struct("<I I I B B 32s")


class NovaFormatError(Exception):
    pass


def pack_header(block_count: int, merkle_root: bytes) -> bytes:
    return HEADER_STRUCT.pack(MAGIC, VERSION, block_count, merkle_root)


def unpack_header(raw: bytes) -> tuple[int, bytes]:
    if len(raw) != HEADER_STRUCT.size:
        raise NovaFormatError("Invalid NOVA header size")
    magic, version, block_count, merkle_root = HEADER_STRUCT.unpack(raw)
    if magic != MAGIC:
        raise NovaFormatError("Invalid NOVA magic")
    if version != VERSION:
        raise NovaFormatError(f"Unsupported NOVA version {version}")
    return block_count, merkle_root


def serialize_block(block: Block) -> bytes:
    header = BLOCK_HEADER_STRUCT.pack(
        block.metadata.block_id,
        block.metadata.uncompressed_size,
        block.metadata.compressed_size,
        block.metadata.compression_method,
        block.metadata.pattern_model,
        block.metadata.hash_bytes,
    )
    return header + block.data


def iter_blocks_from_stream(stream) -> Iterator[Block]:
    while True:
        header_bytes = stream.read(BLOCK_HEADER_STRUCT.size)
        if not header_bytes:
            return
        if len(header_bytes) != BLOCK_HEADER_STRUCT.size:
            raise NovaFormatError("Unexpected truncated block header")
        block_id, uncompressed_size, compressed_size, compression_method, pattern_model, hash_bytes = BLOCK_HEADER_STRUCT.unpack(
            header_bytes
        )
        data = stream.read(compressed_size)
        if len(data) != compressed_size:
            raise NovaFormatError("Unexpected truncated block payload")
        metadata = BlockMetadata(
            block_id=block_id,
            uncompressed_size=uncompressed_size,
            compressed_size=compressed_size,
            compression_method=compression_method,
            pattern_model=pattern_model,
            hash_bytes=hash_bytes,
        )
        yield Block(metadata=metadata, data=data)


def write_nova_file(path, blocks: Iterable[Block], merkle_root: bytes) -> None:
    blocks_list = list(blocks)
    with open(path, "wb") as handle:
        handle.write(pack_header(len(blocks_list), merkle_root))
        for block in blocks_list:
            handle.write(serialize_block(block))


def read_nova_file(path) -> tuple[list[Block], bytes]:
    with open(path, "rb") as handle:
        header = handle.read(HEADER_STRUCT.size)
        block_count, merkle_root = unpack_header(header)
        blocks = list(iter_blocks_from_stream(handle))
        if len(blocks) != block_count:
            raise NovaFormatError("Block count does not match header")
        return blocks, merkle_root

File: novagraph-rust/src/core/format.rs
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::Path;

use crate::core::block::{Block, BlockMetadata};
use crate::crypto::merkle::{compute_merkle_root, sha256};

const NOVA_MAGIC: &[u8; 4] = b"NOVA";
const NOVA_VERSION: u8 = 1;

/// Writes a NOVA container to disk.
pub fn write_nova_file(path: &Path, blocks: &[Block]) -> io::Result<()> {
    let mut file = File::create(path)?;
    file.write_all(NOVA_MAGIC)?;
    file.write_all(&[NOVA_VERSION])?;
    file.write_all(&(blocks.len() as u32).to_le_bytes())?;

    let checksums: Vec<[u8; 32]> = blocks.iter().map(|block| block.metadata.checksum).collect();
    let merkle_root = compute_merkle_root(&checksums);
    file.write_all(&merkle_root)?;

    for block in blocks {
        let bytes = block.metadata.to_bytes();
        file.write_all(&bytes)?;
        file.write_all(&block.payload)?;
    }

    Ok(())
}

/// Reads a NOVA container from disk.
pub fn read_nova_file(path: &Path) -> io::Result<(Vec<Block>, [u8; 32])> {
    let mut file = File::open(path)?;
    let mut header = [0u8; 9];
    file.read_exact(&mut header)?;
    if &header[0..4] != NOVA_MAGIC {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "Invalid NOVA magic"));
    }

    let version = header[4];
    if version != NOVA_VERSION {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "Unsupported NOVA version"));
    }

    let block_count = u32::from_le_bytes(header[5..9].try_into().unwrap());
    let mut merkle_root = [0u8; 32];
    file.read_exact(&mut merkle_root)?;

    let mut blocks = Vec::with_capacity(block_count as usize);
    for _ in 0..block_count {
        let mut metadata_bytes = [0u8; BlockMetadata::SERIALIZED_LEN];
        file.read_exact(&mut metadata_bytes)?;

        let metadata = BlockMetadata::from_bytes(&metadata_bytes)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "Invalid block metadata"))?;

        let mut payload = vec![0u8; metadata.compressed_size as usize];
        file.read_exact(&mut payload)?;

        blocks.push(Block { metadata, payload });
    }

    let computed_root = compute_merkle_root(&blocks.iter().map(|block| block.metadata.checksum).collect::<Vec<_>>());
    if computed_root != merkle_root {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "Merkle root mismatch"));
    }

    Ok((blocks, merkle_root))
}
